modules/platforms/cpp/client/src/ignite_client.cpp

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <thread>
+
+#include "common/ignite_error.h"
+#include "ignite/ignite_client.h"
+#include "ignite_client_impl.h"
+
+namespace ignite
+{
+
+void IgniteClient::startAsync(IgniteClientConfiguration configuration, std::chrono::milliseconds timeout,
+    IgniteCallback<IgniteClient> callback)
+{
+    (void) std::async([cfg = std::move(configuration), timeout, callback = std::move(callback)]() mutable {
+        auto res = IgniteResult<IgniteClient>::ofOperation([cfg = std::move(cfg), timeout] () {
+            return start(cfg, timeout);
```

Comment:
`startAsync` should not call `start`, it should be the other way, because startup is inherently asynchronous.

Reply:
Example:
```cpp
void IgniteClient::startAsync(IgniteClientConfiguration configuration, std::chrono::milliseconds timeout,
    IgniteCallback<IgniteClient> callback)
{
    auto impl = std::make_shared<detail::IgniteClientImpl>(std::move(configuration));

    impl->start([impl, callback] (IgniteResult<void> res) mutable {
        if (!res) {
            // TODO: Extend callback to accept errors somehow
            impl->stop();
        }
        else {
            callback(IgniteClient(std::move(impl)));
        }
    });
}
```

Reply:
Well, ideally yes, but as I already explained before, we do not have asynchronous timers right now to implement timeout properly, so I'm going to leave it as it is for now. I've added a TODO with reference to a ticket though.